### PR TITLE
7051 Universal Pipeline Quality Filtering Results

### PR DESCRIPTION
**Pull request by TurkeyFried**
This PR makes various updates to ensure that the data that you fetch from the History API displays correctly through the Quality filtering step for the Universal Pipeline.

Test Steps:
1. On your `organizations.yml` file, make a change on a `qualityFilter:` entry that you can send reports to. Note: This PR enables `ignore-FULL_ERL` as a receiver for testing purposes. The quality filter section for this receiver also includes a commented out example of a filter that will always trigger.
2. Submit a Full ELR report via Postman (or whichever you use)
3. After receiving a `submissionId`, use that to check the History API until the response includes a list of `destinations`.
4. On receiving a response, check that `itemCount` is lower than `itemCountBeforeQualityFiltering` and that there's data on `filteredReportRows` and `filteredReportItems`
5. Revert the changes on Step 1, then repeat steps 2-4. This time `itemCount` should match `itemCountBeforeQualityFiltering`, and `filteredReportRows` and `filteredReportItems` should be empty.

## Changes
- Added `ignore-FULL_ELR` receiver with filters for testing
- Added logging for the Universal Pipeline during the `route` step
- Updated `SubmissionHistory` to handle the different structure of the translation steps on the pipeline process.

## Linked Issues
- #7051

## To Be Done
- #7052

### CODE REVIEW

File: prime-router/metadata/hl7_mapping/STLTs/ignore/ignore.yml [hunk 1]
```diff
@@ -0,0 +1,14 @@
+hl7Type: ORU_R01
+hl7Version: 2.5.1
+extends: ../../ORU_R01/ORU_R01-base
```
Comment (carlosfelix2) [inline prime-router/metadata/hl7_mapping/STLTs/ignore/ignore.yml:3]: Is this really necessary?  This is not a schema we would want in production.

File: prime-router/settings/organizations.yml [hunk 2]
```diff
@@ -2418,6 +2418,14 @@
   filters:
     - topic: covid-19
       jurisdictionalFilter: [ "matches(ordering_facility_state, IG)" ]
+    - topic: full-elr
+      jurisdictionalFilter:
+        - "(Bundle.entry.resource.ofType(ServiceRequest)[0].requester.resolve().organization.resolve().address.state.exists() and Bundle.entry.resource.ofType(ServiceRequest)[0].requester.resolve().organization.resolve().address.state = 'IG') or (Bundle.entry.resource.ofType(Patient).address.exists() and Bundle.entry.resource.ofType(Patient).address.state = 'IG')"
+#        - "(%performerState.exists() and %performerState = 'IG') or (%patientState.exists() and %patientState = 'IG')"
```
Comment (carlosfelix2) [inline prime-router/settings/organizations.yml:2424]: Remove comment

File: prime-router/settings/organizations.yml [hunk 3]
```diff
@@ -2418,6 +2418,14 @@
   filters:
     - topic: covid-19
       jurisdictionalFilter: [ "matches(ordering_facility_state, IG)" ]
+    - topic: full-elr
+      jurisdictionalFilter:
+        - "(Bundle.entry.resource.ofType(ServiceRequest)[0].requester.resolve().organization.resolve().address.state.exists() and Bundle.entry.resource.ofType(ServiceRequest)[0].requester.resolve().organization.resolve().address.state = 'IG') or (Bundle.entry.resource.ofType(Patient).address.exists() and Bundle.entry.resource.ofType(Patient).address.state = 'IG')"
```
Comment (carlosfelix2) [inline prime-router/settings/organizations.yml:2423]: I think this can use the same setup as the CO filters to make them more readable with %patientState, etc.

File: prime-router/settings/organizations.yml [hunk 4]
```diff
@@ -2565,7 +2573,7 @@
         initialTime: 00:00
       translation:
         type: CUSTOM
-        schemaName: none
+        schemaName: "metadata/hl7_mapping/STLTs/ignore/ignore"
```
Comment (carlosfelix2) [inline prime-router/settings/organizations.yml:2576]: Set this schema to the ORU_R01 and remove the IGNORE schema

File: prime-router/src/main/kotlin/fhirengine/engine/FHIRRouter.kt [hunk 5]
```diff
@@ -379,12 +417,27 @@ class FHIRRouter(
             defaultResponse
         } else {
             filter.all {
-                FhirPathUtils.evaluateCondition(
+                val longhand = replaceShorthand(it)
```
Comment (carlosfelix2) [inline prime-router/src/main/kotlin/fhirengine/engine/FHIRRouter.kt:420]: This val's name is confusing.  Use a more descriptive name

File: prime-router/src/main/kotlin/fhirengine/engine/FHIRRouter.kt [hunk 6]
```diff
@@ -379,12 +417,27 @@ class FHIRRouter(
...
-                    replaceShorthand(it)
+                    longhand
                 )
+                // log results of filtering things OUT
+                if (!evalResult && receiverName != null && filterType != ReportStreamFilterType.JURISDICTIONAL_FILTER) {
+                    report.filteringResults.add(
+                        ReportStreamFilterResult(
+                            receiverName,
+                            report.itemCount,
+                            longhand,
```
Comment (carlosfelix2) [inline prime-router/src/main/kotlin/fhirengine/engine/FHIRRouter.kt:433]: We may want to log the shorthand here.

File: prime-router/src/main/kotlin/history/SubmissionHistory.kt [hunk 7]
```diff
@@ -287,12 +288,21 @@ class DetailedSubmissionHistory(
...
         // Enforce an order on the enrichment:  process/translate, send, download
-        descendants.filter {
-            it.actionName == TaskAction.process || it.actionName == TaskAction.translate
-        }.forEach { descendant ->
-            enrichWithProcessAction(descendant)
+        if (topic == "full-elr") {
+            // logs and destinations are handled very differently for UP
+            // both routing and translate are populated at different times,
+            // so we need to do special logic to handle them
+            enrichWithRoutingAndTranslationActions(descendants)
```
Comment (carlosfelix2) [inline prime-router/src/main/kotlin/history/SubmissionHistory.kt:297]: If this is specific to UP, is it worth including the name of the pipeline here? Is it worth it?

File: prime-router/src/main/kotlin/history/SubmissionHistory.kt [hunk 8]
```diff
@@ -303,17 +313,71 @@ class DetailedSubmissionHistory(
...
+     * Enrich a parent detailed history with details from the route and translate actions.
+     * Add destinations, errors, and warnings, to the history details.
+     * Note: Route/Translate is exclusive to the Universal pipeline
+     * See enrichWithProcessAction for the COVID pipeline counterpart
+     *
+     * @param descendants[] route and translate actions that will be used to enrich
+     */
+    private fun enrichWithRoutingAndTranslationActions(descendants: List<DetailedSubmissionHistory>) {
+        require(topic == "full-elr") {
+            "Must be route or translate actions for a full-elr submission"
```
Comment (carlosfelix2) [inline prime-router/src/main/kotlin/history/SubmissionHistory.kt:325]: This message is confusing as you are checking just for the topic and not the actions.  Replace.

File: prime-router/src/main/kotlin/history/SubmissionHistory.kt [hunk 9]
```diff
@@ -303,17 +313,71 @@ class DetailedSubmissionHistory(
...
+        }
+
+        val filterLogs = mutableListOf<DetailedActionLog>()
+
+        descendants.filter { it.actionName == TaskAction.route }.forEach { descendant ->
```
Comment (carlosfelix2) [inline prime-router/src/main/kotlin/history/SubmissionHistory.kt:330]: Add some more comments, so we can understand this

**Review by carlosfelix2 — CHANGES_REQUESTED** [review]

File: prime-router/metadata/hl7_mapping/STLTs/ignore/ignore.yml (hunk 1 above)
Comment (TurkeyFried) [reply prime-router/metadata/hl7_mapping/STLTs/ignore/ignore.yml]: Done, didn't realize this was one of the prod-facing files!

File: prime-router/settings/organizations.yml (hunk 2 above)
Comment (TurkeyFried) [reply prime-router/settings/organizations.yml]: fixed

File: prime-router/settings/organizations.yml (hunk 3 above)
Comment (TurkeyFried) [reply prime-router/settings/organizations.yml]: Yup, looks a lot better after doing that

File: prime-router/settings/organizations.yml (hunk 4 above)
Comment (TurkeyFried) [reply prime-router/settings/organizations.yml]: done

File: prime-router/src/main/kotlin/fhirengine/engine/FHIRRouter.kt (hunk 5 above)
Comment (TurkeyFried) [reply prime-router/src/main/kotlin/fhirengine/engine/FHIRRouter.kt]: i like the name!
fixed tho, and another tweak made the val not needed at all

File: prime-router/src/main/kotlin/fhirengine/engine/FHIRRouter.kt (hunk 6 above)
Comment (TurkeyFried) [reply prime-router/src/main/kotlin/fhirengine/engine/FHIRRouter.kt]: Yup that looks a lot better

File: prime-router/src/main/kotlin/history/SubmissionHistory.kt (hunk 8 above)
Comment (TurkeyFried) [reply prime-router/src/main/kotlin/history/SubmissionHistory.kt]: updated

File: prime-router/src/main/kotlin/fhirengine/engine/FHIRRouter.kt [hunk 10]
```diff
@@ -371,6 +406,9 @@ class FHIRRouter(
         filter: ReportStreamFilter?,
         bundle: Bundle,
         defaultResponse: Boolean,
+        report: Report,
```
Comment (clediggins-usds) [inline prime-router/src/main/kotlin/fhirengine/engine/FHIRRouter.kt:409]: I don't think we want the `report` in this function. Tracking the results should be done outside of the 'evaluateFilterCondition' because you have now made this 'evaluateFilterConditionAndTrackResults'

**Review by clediggins-usds — CHANGES_REQUESTED** [review]

File: prime-router/settings/organizations.yml [hunk 11]
```diff
@@ -2418,6 +2418,11 @@
   filters:
     - topic: covid-19
       jurisdictionalFilter: [ "matches(ordering_facility_state, IG)" ]
+    - topic: full-elr
+      jurisdictionalFilter:
+        - "%performerState = 'IG' or %patientState = 'IG'"
+      qualityFilter:
```
[the remark this answers is not in the document]
Comment (TurkeyFried) [reply prime-router/settings/organizations.yml:2424]: I added it to make it easier to make quality filter changes that filter things out. For example, changing R01 to R02 there will 100% trigger a filtering out without breaking anything.

File: prime-router/metadata/tables/local/fhirpath_filter_shorthand.csv [hunk 12]
```diff
@@ -6,4 +6,6 @@ performerState,Bundle.entry.resource.ofType(ServiceRequest)[0].requester.resolve
 patientState,Bundle.entry.resource.ofType(Patient).address.state
 specimen,Bundle.entry.resource.ofType(Specimen)
 serviceRequest,Bundle.entry.resource.ofType(ServiceRequest)
-observation,Bundle.entry.resource.ofType(Observation)
\ No newline at end of file
+observation,Bundle.entry.resource.ofType(Observation)
+provenance,Bundle.entry.resource.ofType(Provenance)
+codeSystem,Bundle.entry.resource.ofType(Provenance).activity.coding.code
```
[the remark this answers is not in the document]
Comment (TurkeyFried) [reply prime-router/metadata/tables/local/fhirpath_filter_shorthand.csv]: yeah that's a much better name, done!

File: prime-router/src/main/kotlin/fhirengine/engine/FHIRRouter.kt (hunk 10 above)
Comment (TurkeyFried) [reply prime-router/src/main/kotlin/fhirengine/engine/FHIRRouter.kt]: cleaned the parameters up on that function while taking out the logging from it

File: prime-router/src/main/kotlin/fhirengine/engine/FHIRRouter.kt [hunk 13]
```diff
@@ -327,29 +315,76 @@ class FHIRRouter(
 
             // JURIS FILTER
             //  default: allowNone
-            var passes = evaluateFilterCondition(getJurisFilters(receiver, orgFilters), bundle, false)
+            var passes = evaluateFilterCondition(
+                getJurisFilters(receiver, orgFilters),
+                bundle,
+                false,
+                report,
```
[the remark this answers is not in the document]
Comment (TurkeyFried) [reply prime-router/src/main/kotlin/fhirengine/engine/FHIRRouter.kt]: done

File: prime-router/src/main/kotlin/history/SubmissionHistory.kt [hunk 14]
```diff
@@ -287,12 +288,21 @@ class DetailedSubmissionHistory(
      */
     fun enrichWithDescendants(descendants: List<DetailedSubmissionHistory>) {
         check(descendants.distinctBy { it.actionId }.size == descendants.size)
+
         // Enforce an order on the enrichment:  process/translate, send, download
-        descendants.filter {
-            it.actionName == TaskAction.process || it.actionName == TaskAction.translate
-        }.forEach { descendant ->
-            enrichWithProcessAction(descendant)
+        if (topic == "full-elr") {
```
[the remark this answers is not in the document]
Comment (TurkeyFried) [reply prime-router/src/main/kotlin/history/SubmissionHistory.kt]: good catch, fixed

File: prime-router/src/main/kotlin/fhirengine/engine/FHIRRouter.kt [hunk 15]
```diff
@@ -379,12 +417,26 @@ class FHIRRouter(
...
             filter.all {
-                FhirPathUtils.evaluateCondition(
+                val evalResult = FhirPathUtils.evaluateCondition(
                     CustomContext(bundle, bundle, shorthandLookupTable),
                     bundle,
                     bundle,
                     replaceShorthand(it)
                 )
+                // log results of filtering things OUT
+                if (!evalResult && receiverName != null && filterType != ReportStreamFilterType.JURISDICTIONAL_FILTER) {
```
[the remark this answers is not in the document]
Comment (TurkeyFried) [reply prime-router/src/main/kotlin/fhirengine/engine/FHIRRouter.kt]: fixed

File: prime-router/src/main/kotlin/history/SubmissionHistory.kt (hunk 9 above)
Comment (TurkeyFried) [reply prime-router/src/main/kotlin/history/SubmissionHistory.kt:332]: done

File: prime-router/src/main/kotlin/history/SubmissionHistory.kt (hunk 7 above)
Comment (TurkeyFried) [reply prime-router/src/main/kotlin/history/SubmissionHistory.kt:298]: Thinking of a good name....anyone who sees this, feel free to suggest ideas

Comment (jimmyfagan) [reply prime-router/src/main/kotlin/history/SubmissionHistory.kt:298]: My naive opinion here would be that we don't need it in the name (the name is pretty long to begin with). The comment here already calls out that it is specific to UP, and the function comments for `enrichWithRoutingAndTranslationActions` and `enrichWithProcessAction` already mention that they are specific to one or the other pipeline.

File: prime-router/src/main/kotlin/fhirengine/engine/FHIRRouter.kt [hunk 16]
```diff
@@ -328,29 +316,62 @@ class FHIRRouter(
...
-            //           must have at least one of order test date, specimen collection date/time, test result date
-            passes = passes &&
-                evaluateFilterCondition(
-                    getQualityFilters(receiver, orgFilters),
+            if (passes) {
+                //  default: must have message id, patient last name, patient first name, dob, specimen type
+                //           must have at least one of patient street, zip code, phone number, email
+                //           must have at least one of order test date, specimen collection date/time, test result date
+                // get the quality filter default result for the bundle, but only if it is needed
+                val qualFilterDefaultResult: Boolean by lazy {
```
Comment (clediggins-usds) [inline prime-router/src/main/kotlin/fhirengine/engine/FHIRRouter.kt:326]: Why did you move this? The default value is one-per-bundle, regardless of which receiver it is for, and should not be taking 'receiver. resverTheQualityFilter into account since this would result in a double negation - it would negate the default and then negate that negation when it ran the evaluation.

File: prime-router/src/main/kotlin/fhirengine/engine/FHIRRouter.kt [hunk 17]
```diff
@@ -328,29 +316,62 @@ class FHIRRouter(
...
-            passes = passes &&
-                evaluateFilterCondition(
-                    getProcessingModeFilter(receiver, orgFilters),
-                    bundle,
-                    processingModeDefaultResult
-                )
+            if (passes) {
+                //  default: allowAll
+                // get the processing mode (processing id) default result for the bundle, but only if it is needed
+                val processingModeDefaultResult: Boolean by lazy {
```
Comment (clediggins-usds) [inline prime-router/src/main/kotlin/fhirengine/engine/FHIRRouter.kt:363]: Why did you move this?  It is one-per-bundle, no need to calculate for each receiver that passes the previous filters.

File: prime-router/src/main/kotlin/fhirengine/engine/FHIRRouter.kt [hunk 18]
```diff
@@ -377,19 +398,37 @@ class FHIRRouter(
...
-        } else {
-            filter.all {
-                FhirPathUtils.evaluateCondition(
-                    CustomContext(bundle, bundle, shorthandLookupTable),
-                    bundle,
-                    bundle,
-                    replaceShorthand(it)
-                )
-            }
+        } else filter.all {
```
Comment (clediggins-usds) [inline prime-router/src/main/kotlin/fhirengine/engine/FHIRRouter.kt:401]: Please add the {} back in for readability around the 'else' block for consistency.

File: prime-router/src/main/kotlin/fhirengine/engine/FHIRRouter.kt [hunk 19]
```diff
@@ -377,19 +398,37 @@ class FHIRRouter(
...
+                CustomContext(bundle, bundle, shorthandLookupTable),
+                bundle,
+                bundle,
+                replaceShorthand(it)
+            )
         }
         return if (reverseFilter) !result else result
     }
 
+    internal fun logFilterResults(
```
Comment (clediggins-usds) [inline prime-router/src/main/kotlin/fhirengine/engine/FHIRRouter.kt:412]: kdoc please

File: prime-router/src/test/kotlin/fhirengine/engine/fhirRouterTests/RoutingTests.kt [hunk 20]
```diff
@@ -623,8 +641,27 @@ class RoutingTests {
...
+        val bundle = FhirTranscoder.decode(fhirData)
+        val report = Report(one, listOf(listOf("1", "2")), TestSource, metadata = UnitTestUtils.simpleMetadata)
+        val settings = FileSettings().loadOrganizations(oneOrganization)
+        var qualFilter = listOf("Bundle.entry.resource.ofType(Provenance).count() > 0")
+
+        val engine = spyk(makeFhirEngine(metadata, settings, TaskAction.route) as FHIRRouter)
+
+        assertThat(report.filteringResults.count()).isEqualTo(0)
+        engine.logFilterResults(qualFilter, bundle, report, receiver, ReportStreamFilterType.QUALITY_FILTER)
+        assertThat(report.filteringResults.count()).isGreaterThan(0)
```
Comment (clediggins-usds) [inline prime-router/src/test/kotlin/fhirengine/engine/fhirRouterTests/RoutingTests.kt:665]: Add an assertThat that actually looks at what is inside the filteringResults value and verifies the correct (and useful) data is present.

**Review by clediggins-usds — CHANGES_REQUESTED** [review]

File: prime-router/src/main/kotlin/fhirengine/engine/FHIRRouter.kt (hunk 17 above)
Comment (TurkeyFried) [reply prime-router/src/main/kotlin/fhirengine/engine/FHIRRouter.kt:363]: I moved it just to accompany the other one, which I thought needed to be moved but didn't actually upon review. fixing.....

File: prime-router/src/main/kotlin/fhirengine/engine/FHIRRouter.kt (hunk 16 above)
Comment (TurkeyFried) [reply prime-router/src/main/kotlin/fhirengine/engine/FHIRRouter.kt]: I had done that to handle the logging before understanding it better. I've reverted those changes.

File: prime-router/src/main/kotlin/fhirengine/engine/FHIRRouter.kt [hunk 21]
```diff
@@ -288,17 +290,19 @@ class FHIRRouter(
 
     /**
      * Applies all filters to the list of all receivers with topic FULL_ELR that are not set as INACTIVE.
-     * FHIRPath expressions are run against the [bundle] to determine if the receiver should get this message\
+     * FHIRPath expressions are run against the [bundle] to determine if the receiver should get this message
+     * @param bundle
+     * @param report
```
Comment (jimmyfagan) [inline prime-router/src/main/kotlin/fhirengine/engine/FHIRRouter.kt:295]: Some feedback I got from Carlos on another PR is that param descriptions should all be inline with the function description, like what is already there for `[bundle]`. I think that means `report` should be documented the same way, and we don't need `@param bundle` or `@param report`.

File: prime-router/src/main/kotlin/fhirengine/engine/FHIRRouter.kt (hunk 19 above)
Comment (TurkeyFried) [reply prime-router/src/main/kotlin/fhirengine/engine/FHIRRouter.kt:427]: done

File: prime-router/src/main/kotlin/fhirengine/engine/FHIRRouter.kt (hunk 18 above)
Comment (TurkeyFried) [reply prime-router/src/main/kotlin/fhirengine/engine/FHIRRouter.kt]: done

File: prime-router/src/test/kotlin/fhirengine/engine/fhirRouterTests/RoutingTests.kt (hunk 20 above)
Comment (TurkeyFried) [reply prime-router/src/test/kotlin/fhirengine/engine/fhirRouterTests/RoutingTests.kt]: Makes sense. done!

File: prime-router/src/main/kotlin/history/SubmissionHistory.kt [hunk 22]
```diff
@@ -303,17 +314,73 @@ class DetailedSubmissionHistory(
     }
 
     /**
-     * Enrich a parent detailed history with details from the process or translate action.
+     * Enrich a parent detailed history with details from the route and translate actions.
+     * Add destinations, errors, and warnings, to the history details.
+     * Note: Route/Translate is exclusive to the Universal pipeline
+     * See enrichWithProcessAction for the COVID pipeline counterpart
```
[the remark this answers is not in the document]
Comment (TurkeyFried) [reply prime-router/src/main/kotlin/history/SubmissionHistory.kt:320]: I'm so used to hearing COVID pipeline that I've been taking that name for granted. Need to check if we have an official name for it...

File: prime-router/src/main/kotlin/fhirengine/engine/FHIRRouter.kt (hunk 21 above)
Comment (TurkeyFried) [reply prime-router/src/main/kotlin/fhirengine/engine/FHIRRouter.kt]: Fixed

File: prime-router/src/main/kotlin/history/SubmissionHistory.kt (hunk 22 above)
[the remark this answers is not in the document]
Comment (clediggins-usds) [reply prime-router/src/main/kotlin/history/SubmissionHistory.kt:320]: Updating this to read 'TopicReceiver pipeline' instead of 'covid pipeline' would be more accurate, but generally everyone calls it the 'covid pipeline' so I think the message is understood.

File: prime-router/src/test/kotlin/fhirengine/engine/fhirRouterTests/RoutingTests.kt [hunk 23]
```diff
@@ -2,6 +2,7 @@ package gov.cdc.prime.router.fhirengine.engine.fhirRouterTests
 
 import assertk.assertThat
 import assertk.assertions.isEqualTo
+import assertk.assertions.isGreaterThan
```
Comment (jimmyfagan) [inline prime-router/src/test/kotlin/fhirengine/engine/fhirRouterTests/RoutingTests.kt:5]: Just noticed that this is an unused import, which make ktlint complain, try to fix this before merging!

File: prime-router/src/main/kotlin/history/SubmissionHistory.kt [hunk 24]
```diff
@@ -303,17 +314,73 @@ class DetailedSubmissionHistory(
...
+
+        // Grab the filter logs generated during the "route" action, as well as errors and warnings
+        descendants.filter { it.actionName == TaskAction.route }.forEach { descendant ->
+            filterLogs += descendant.logs.filter { log -> log.type == ActionLogLevel.filter }
+            errors += descendant.errors
+            warnings += descendant.warnings
+        }
+
+        // Grab destinations from the "translate" action, if the submission made it that far
+        descendants.filter { it.actionName == TaskAction.translate }.forEach { descendant  ->
```
Comment (jimmyfagan) [inline prime-router/src/main/kotlin/history/SubmissionHistory.kt:339]: The linter doesn't like this line either, you have an extra " " between `descendant` and `->`,  try to fix that before merging as well!

File: prime-router/src/main/kotlin/history/SubmissionHistory.kt (hunk 7 above)
Comment (TurkeyFried) [reply prime-router/src/main/kotlin/history/SubmissionHistory.kt:298]: Will leave as is for now

File: prime-router/src/main/kotlin/history/SubmissionHistory.kt (hunk 22 above)
[the remark this answers is not in the document]
Comment (TurkeyFried) [reply prime-router/src/main/kotlin/history/SubmissionHistory.kt]: Updated the name just to get started on the habit

File: prime-router/src/main/kotlin/history/SubmissionHistory.kt (hunk 24 above)
Comment (TurkeyFried) [reply prime-router/src/main/kotlin/history/SubmissionHistory.kt]: Yup, forgot to run lint. Reran and it's fixed now

File: prime-router/src/test/kotlin/fhirengine/engine/fhirRouterTests/RoutingTests.kt (hunk 23 above)
Comment (TurkeyFried) [reply prime-router/src/test/kotlin/fhirengine/engine/fhirRouterTests/RoutingTests.kt]: fixed

File: prime-router/settings/organizations.yml [hunk 25]
```diff
@@ -2418,6 +2418,11 @@
...
+        - "%hl7MessageType = 'R01'"
```
Comment (carlosfelix2) [inline prime-router/settings/organizations.yml:2425]: This is fine as a test here, but just FYI that this is not a proper quality filter for a receiver.  Just think that the system is ingesting all sorts of data and this receiver is getting information that a bunch of data was filtered just because it was not an observation.

File: prime-router/src/main/kotlin/fhirengine/engine/FHIRRouter.kt [hunk 26]
```diff
@@ -328,29 +332,49 @@ class FHIRRouter(
...
             var passes = evaluateFilterCondition(getJurisFilters(receiver, orgFilters), bundle, false)
+
             // QUALITY FILTER
-            //  default: must have message id, patient last name, patient first name, dob, specimen type
-            //           must have at least one of patient street, zip code, phone number, email
-            //           must have at least one of order test date, specimen collection date/time, test result date
-            passes = passes &&
-                evaluateFilterCondition(
-                    getQualityFilters(receiver, orgFilters),
+            if (passes) {
```
Comment (carlosfelix2) [inline prime-router/src/main/kotlin/fhirengine/engine/FHIRRouter.kt:337]: FYI that this is functionally the same as the passes &&

Comment (carlosfelix2) [inline prime-router/src/main/kotlin/fhirengine/engine/FHIRRouter.kt:337]: Rename passes to something more meaningful since it is now outside of the passes &&.  notFiltered?

File: prime-router/src/main/kotlin/fhirengine/engine/FHIRRouter.kt [hunk 27]
```diff
@@ -328,29 +332,49 @@ class FHIRRouter(
...
+                //  default: must have message id, patient last name, patient first name, dob, specimen type
+                //           must have at least one of patient street, zip code, phone number, email
+                //           must have at least one of order test date, specimen collection date/time, test result date
+                val filters = getQualityFilters(receiver, orgFilters)
```
Comment (carlosfelix2) [inline prime-router/src/main/kotlin/fhirengine/engine/FHIRRouter.kt:341]: Remove this val and put this code online.  Kotlin is about reducing unneeded lines of code.

File: prime-router/src/main/kotlin/fhirengine/engine/FHIRRouter.kt [hunk 28]
```diff
@@ -390,6 +414,34 @@ class FHIRRouter(
         return if (reverseFilter) !result else result
```
Comment (carlosfelix2) [inline prime-router/src/main/kotlin/fhirengine/engine/FHIRRouter.kt:414]: OMG! I just noticed this. This propery is `reverseTheQualityFilter` which means it only applies to the quality filter.  You can open a new ticket for this if you want.

File: prime-router/src/main/kotlin/fhirengine/engine/FHIRRouter.kt [hunk 29]
```diff
@@ -390,6 +414,34 @@ class FHIRRouter(
...
     }
 
+    /**
+     * Log the results of filtering items out of reports during the "route" step
+     *
+     * @param filters - Filters run on the items
```
Comment (carlosfelix2) [inline prime-router/src/main/kotlin/fhirengine/engine/FHIRRouter.kt:420]: Change this kdoc to have the parameters online.  This is what the program wanted.

File: prime-router/src/main/kotlin/history/SubmissionHistory.kt [hunk 30]
```diff
@@ -303,17 +314,73 @@ class DetailedSubmissionHistory(
...
+        // these dummies are created using the receivers defined in the logs
+        if (destinations.isEmpty() && filterLogs.isNotEmpty()) {
+            val filteredReportRows = filterLogs.map { log -> log.detail.message }
+            val filteredReportItems = filterLogs.map { log ->
+                ReportStreamFilterResultForResponse(log.detail as ReportStreamFilterResult)
+            }
+            filterLogs.forEach { log ->
+                val casted = log.detail as ReportStreamFilterResult
+
+                val splits = casted.receiverName.split(Sender.fullNameSeparator)
```
Comment (carlosfelix2) [inline prime-router/src/main/kotlin/history/SubmissionHistory.kt:356]: Better val name

File: prime-router/src/main/kotlin/history/SubmissionHistory.kt [hunk 31]
```diff
@@ -303,17 +314,73 @@ class DetailedSubmissionHistory(
...
+        // if all the items were filtered out, no translate actions are generated (thus no destinations)
+        // to return the expected output, we create dummy destinations that are only used in enrichment
+        // these dummies are created using the receivers defined in the logs
+        if (destinations.isEmpty() && filterLogs.isNotEmpty()) {
+            val filteredReportRows = filterLogs.map { log -> log.detail.message }
+            val filteredReportItems = filterLogs.map { log ->
+                ReportStreamFilterResultForResponse(log.detail as ReportStreamFilterResult)
+            }
+            filterLogs.forEach { log ->
+                val casted = log.detail as ReportStreamFilterResult
```
Comment (carlosfelix2) [inline prime-router/src/main/kotlin/history/SubmissionHistory.kt:354]: Cast this inline.  Reduce unneeded lines.

**Review by carlosfelix2 — APPROVED** [review]
with comments

File: prime-router/src/main/kotlin/fhirengine/engine/FHIRRouter.kt (hunk 28 above)
Comment (TurkeyFried) [reply prime-router/src/main/kotlin/fhirengine/engine/FHIRRouter.kt:414]: Ticket written!

File: prime-router/src/main/kotlin/fhirengine/engine/FHIRRouter.kt (hunk 26 above)
Comment (TurkeyFried) [reply prime-router/src/main/kotlin/fhirengine/engine/FHIRRouter.kt]: updated

File: prime-router/src/main/kotlin/fhirengine/engine/FHIRRouter.kt (hunk 27 above)
Comment (TurkeyFried) [reply prime-router/src/main/kotlin/fhirengine/engine/FHIRRouter.kt]: the val is used twice, but with the other changes I put in the code change is not much

File: prime-router/src/main/kotlin/fhirengine/engine/FHIRRouter.kt (hunk 29 above)
Comment (TurkeyFried) [reply prime-router/src/main/kotlin/fhirengine/engine/FHIRRouter.kt]: Done

File: prime-router/src/main/kotlin/history/SubmissionHistory.kt (hunk 30 above)
Comment (TurkeyFried) [reply prime-router/src/main/kotlin/history/SubmissionHistory.kt]: done!

**Review by clediggins-usds — APPROVED** [review]
